File: python/mirage/commands/cli/builtin/gh/actions.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
from typing import Any

from mirage.commands.cli.builtin.gh.accessor import (camel, list_limit,
                                                     read_cli_file, repo_for,
                                                     text_out, typed_out)
from mirage.commands.cli.types import CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.core.github.actions import (dispatch_workflow, get_run,
                                        get_workflow, list_runs,
                                        list_workflows, rerun, rerun_job)
from mirage.core.github.config import GhConfig
from mirage.core.github.repo import view_repo
from mirage.io.stream import materialize
from mirage.io.types import ByteSource, IOResult
from mirage.types import JsonValue

RUN_FIELDS = ("attempt", "conclusion", "createdAt", "databaseId",
              "displayTitle", "event", "headBranch", "headSha", "name",
              "number", "startedAt", "status", "updatedAt", "url",
              "workflowDatabaseId", "workflowName")
WORKFLOW_FIELDS = ("id", "name", "path", "state")


def _run(value: JsonValue) -> dict[str, Any]:
    row = camel(value)
    result = row if isinstance(row, dict) else {}
    if "id" in result:
        result["databaseId"] = result.pop("id")
    if "htmlUrl" in result:
        result["url"] = result.pop("htmlUrl")
    if "runAttempt" in result:
        result["attempt"] = result.pop("runAttempt")
    if "runNumber" in result:
        result["number"] = result.pop("runNumber")
    if "workflowId" in result:
        result["workflowDatabaseId"] = result.pop("workflowId")
    result.setdefault("workflowName", result.get("name", ""))
    result.setdefault("startedAt", result.get("runStartedAt"))
    return result


def _workflow(value: JsonValue) -> dict[str, Any]:
    row = camel(value)
    return row if isinstance(row, dict) else {}


async def run_list_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    params: dict[str, str] = {}
    for flag, key in (("branch", "branch"), ("commit", "head_sha"),
                      ("event", "event"), ("status", "status"),
                      ("user", "actor"), ("created", "created")):
        value = fl.as_str(flag)
        if value:
            params[key] = value
    rows = [
        _run(value)
        for value in await list_runs(inv.config, repo_for(inv, fl), params,
                                     list_limit(fl, 20), fl.as_str("workflow"))
    ]
    human = "".join(f'{row.get("status", "")}\t'
                    f'{row.get("conclusion", "")}\t'
                    f'{row.get("displayTitle", "")}\t'
                    f'{row.get("workflowName", "")}\t'
                    f'{row.get("headBranch", "")}\t'
                    f'{row.get("event", "")}\t{row.get("databaseId", "")}\n'
                    for row in rows)
    return await typed_out(rows, fl, human, RUN_FIELDS)


async def run_view_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    raw = inv.texts[0] if inv.texts else ""
    if not raw.isdigit():
        raise ValueError("a run ID is required in noninteractive mode")
    row = _run(await get_run(inv.config, repo_for(inv, fl), int(raw)))
    human = (f'title:\t{row.get("displayTitle", "")}\n'
             f'workflow:\t{row.get("workflowName", "")}\n'
             f'status:\t{row.get("status", "")}\n'
             f'conclusion:\t{row.get("conclusion", "")}\n'
             f'branch:\t{row.get("headBranch", "")}\n'
             f'event:\t{row.get("event", "")}\n')
    out, io = await typed_out(row, fl, human, RUN_FIELDS)
    if fl.as_bool("exit_status") and row.get("conclusion") not in (None, "",
                                                                   "success"):
        io.exit_code = 1
    return out, io


async def run_rerun_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    raw = inv.texts[0] if inv.texts else ""
    if not raw.isdigit():
        raise ValueError("a run ID is required in noninteractive mode")
    ref = repo_for(inv, fl)
    job = fl.as_str("job")
    if job:
        if not job.isdigit():
            raise ValueError("--job expects a numeric job ID")
        await rerun_job(inv.config, ref, int(job), fl.as_bool("debug"))
    else:
        suffix = "rerun-failed-jobs" if fl.as_bool("failed") else "rerun"
        body: JsonValue = ({
            "enable_debug_logging": True
        } if fl.as_bool("debug") else None)
        await rerun(inv.config, ref, int(raw), suffix, body)
    return text_out("")


async def workflow_list_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    include = (None if fl.as_bool("all") else
               lambda row: row.get("state") == "active")
    rows = [
        _workflow(value) for value in await list_workflows(
            inv.config, repo_for(inv, fl), list_limit(fl, 50), include=include)
    ]
    human = "".join(f'{row.get("name", "")}\t{row.get("state", "")}\t'
                    f'{row.get("id", "")}\n' for row in rows)
    return await typed_out(rows, fl, human, WORKFLOW_FIELDS)


async def workflow_view_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    workflow = inv.texts[0] if inv.texts else ""
    if not workflow:
        raise ValueError("a workflow ID, name, or filename is required")
    row = _workflow(await get_workflow(inv.config, repo_for(inv, fl),
                                       workflow))
    human = (f'{row.get("name", "")} - {row.get("state", "")}\n'
             f'ID: {row.get("id", "")}\nFile: {row.get("path", "")}\n')
    return text_out(human)


async def _workflow_inputs(inv: CLIInvocation[GhConfig],
                           fl: FlagView) -> dict[str, JsonValue]:
    if fl.as_bool("json"):
        if inv.stdin is None:
            raise ValueError("--json needs standard input")
        try:
            value = json.loads((await materialize(inv.stdin)).decode())
        except json.JSONDecodeError as exc:
            raise ValueError(
                f"invalid JSON from standard input: {exc.msg}") from None
        if not isinstance(value, dict):
            raise ValueError("workflow inputs must be a JSON object")
        return value
    inputs: dict[str, JsonValue] = {}
    for pair in fl.as_list("raw_field"):
        key, sep, value = pair.partition("=")
        if not sep:
            raise ValueError(f'expected "key=value", got "{pair}"')
        inputs[key] = value
    for pair in fl.as_list("field"):
        key, sep, value = pair.partition("=")
        if not sep:
            raise ValueError(f'expected "key=value", got "{pair}"')
        inputs[key] = ((await read_cli_file(inv, value[1:],
                                            "--field")).decode()
                       if value.startswith("@") else value)
    return inputs


async def workflow_run_cmd(
        inv: CLIInvocation[GhConfig]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    workflow = inv.texts[0] if inv.texts else ""
    if not workflow:
        raise ValueError("a workflow ID, name, or filename is required")
    ref = repo_for(inv, fl)
    branch = fl.as_str("ref") or inv.config.branch
    if not branch:
        repo = await view_repo(inv.config, ref)
        candidate = repo.get("default_branch") if isinstance(repo,
                                                             dict) else None
        branch = candidate if isinstance(candidate, str) else None
    if not isinstance(branch, str) or not branch:
        raise ValueError("a workflow ref is required")
    body: dict[str, JsonValue] = {
        "ref": branch,
        "inputs": await _workflow_inputs(inv, fl)
    }
    await dispatch_workflow(inv.config, ref, workflow, body)
    return text_out("")

File: typescript/packages/core/src/commands/cli/builtin/gh/actions.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { FlagView } from '../../../spec/flag_view.ts'
import type { CommandFnResult } from '../../../config.ts'
import type { CLIInvocation } from '../../types.ts'
import {
  dispatchWorkflow,
  getRun,
  getWorkflow,
  listRuns,
  listWorkflows,
  rerun,
  rerunJob,
} from '../../../../core/github/actions.ts'
import { viewRepo } from '../../../../core/github/repo.ts'
import { materialize } from '../../../../io/types.ts'
import {
  camel,
  ghTransport,
  readCliFile,
  repoFor,
  textOut,
  textValue,
  typedOut,
} from './accessor.ts'

const RUN_FIELDS = [
  'attempt',
  'conclusion',
  'createdAt',
  'databaseId',
  'displayTitle',
  'event',
  'headBranch',
  'headSha',
  'name',
  'number',
  'startedAt',
  'status',
  'updatedAt',
  'url',
  'workflowDatabaseId',
  'workflowName',
] as const
const WORKFLOW_FIELDS = ['id', 'name', 'path', 'state'] as const

function run(value: unknown): Record<string, unknown> {
  const row = camel(value)
  const result = row !== null && typeof row === 'object' ? (row as Record<string, unknown>) : {}
  if ('id' in result) {
    result.databaseId = result.id
    delete result.id
  }
  if ('htmlUrl' in result) {
    result.url = result.htmlUrl
    delete result.htmlUrl
  }
  if ('runAttempt' in result) {
    result.attempt = result.runAttempt
    delete result.runAttempt
  }
  if ('runNumber' in result) {
    result.number = result.runNumber
    delete result.runNumber
  }
  if ('workflowId' in result) {
    result.workflowDatabaseId = result.workflowId
    delete result.workflowId
  }
  result.workflowName ??= result.name ?? ''
  result.startedAt ??= result.runStartedAt
  return result
}

function workflow(value: unknown): Record<string, unknown> {
  const row = camel(value)
  return row !== null && typeof row === 'object' ? (row as Record<string, unknown>) : {}
}

export async function runListCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const params: Record<string, string> = {}
  for (const [flag, key] of [
    ['branch', 'branch'],
    ['commit', 'head_sha'],
    ['event', 'event'],
    ['status', 'status'],
    ['user', 'actor'],
    ['created', 'created'],
  ] as const) {
    const value = fl.asStr(flag)
    if (value !== undefined && value !== '') params[key] = value
  }
  const rows = (
    await listRuns(
      ghTransport(inv.config),
      repoFor(inv, fl),
      params,
      fl.asInt('limit') ?? 20,
      fl.asStr('workflow'),
    )
  ).map(run)
  const human = rows
    .map(
      (row) =>
        `${textValue(row.status)}\t${textValue(row.conclusion)}\t${textValue(row.displayTitle)}\t${textValue(row.workflowName)}\t${textValue(row.headBranch)}\t${textValue(row.event)}\t${textValue(row.databaseId)}\n`,
    )
    .join('')
  return typedOut(rows, fl, human, RUN_FIELDS)
}

export async function runViewCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const raw = inv.texts[0] ?? ''
  if (!/^\d+$/.test(raw)) throw new Error('a run ID is required in noninteractive mode')
  const row = run(await getRun(ghTransport(inv.config), repoFor(inv, fl), Number(raw)))
  const human = `title:\t${textValue(row.displayTitle)}\nworkflow:\t${textValue(row.workflowName)}\nstatus:\t${textValue(row.status)}\nconclusion:\t${textValue(row.conclusion)}\nbranch:\t${textValue(row.headBranch)}\nevent:\t${textValue(row.event)}\n`
  const out = await typedOut(row, fl, human, RUN_FIELDS)
  if (
    out !== null &&
    fl.asBool('exit_status') &&
    row.conclusion !== null &&
    row.conclusion !== undefined &&
    row.conclusion !== '' &&
    row.conclusion !== 'success'
  ) {
    out[1].exitCode = 1
  }
  return out
}

export async function runRerunCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const raw = inv.texts[0] ?? ''
  if (!/^\d+$/.test(raw)) throw new Error('a run ID is required in noninteractive mode')
  const transport = ghTransport(inv.config)
  const ref = repoFor(inv, fl)
  const job = fl.asStr('job')
  if (job !== undefined && job !== '') {
    if (!/^\d+$/.test(job)) throw new Error('--job expects a numeric job ID')
    await rerunJob(transport, ref, Number(job), fl.asBool('debug'))
  } else {
    await rerun(
      transport,
      ref,
      Number(raw),
      fl.asBool('failed') ? 'rerun-failed-jobs' : 'rerun',
      fl.asBool('debug') ? { enable_debug_logging: true } : undefined,
    )
  }
  return textOut('')
}

export async function workflowListCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const rows = (
    await listWorkflows(
      ghTransport(inv.config),
      repoFor(inv, fl),
      fl.asInt('limit') ?? 50,
      fl.asBool('all') ? undefined : (row) => row.state === 'active',
    )
  ).map(workflow)
  const human = rows
    .map((row) => `${textValue(row.name)}\t${textValue(row.state)}\t${textValue(row.id)}\n`)
    .join('')
  return typedOut(rows, fl, human, WORKFLOW_FIELDS)
}

export async function workflowViewCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const id = inv.texts[0] ?? ''
  if (id === '') throw new Error('a workflow ID, name, or filename is required')
  const row = workflow(await getWorkflow(ghTransport(inv.config), repoFor(inv, fl), id))
  return textOut(
    `${textValue(row.name)} - ${textValue(row.state)}\nID: ${textValue(row.id)}\nFile: ${textValue(row.path)}\n`,
  )
}

async function workflowInputs(inv: CLIInvocation, fl: FlagView): Promise<Record<string, unknown>> {
  if (fl.asBool('json')) {
    if (inv.stdin === null) throw new Error('--json needs standard input')
    let value: unknown
    try {
      value = JSON.parse(new TextDecoder().decode(await materialize(inv.stdin)))
    } catch (err) {
      if (err instanceof SyntaxError)
        throw new Error(`invalid JSON from standard input: ${err.message}`)
      throw err
    }
    if (value === null || typeof value !== 'object' || Array.isArray(value)) {
      throw new Error('workflow inputs must be a JSON object')
    }
    return value as Record<string, unknown>
  }
  const inputs: Record<string, unknown> = {}
  for (const pair of fl.asList('raw_field')) {
    const at = pair.indexOf('=')
    if (at < 0) throw new Error(`expected "key=value", got "${pair}"`)
    inputs[pair.slice(0, at)] = pair.slice(at + 1)
  }
  for (const pair of fl.asList('field')) {
    const at = pair.indexOf('=')
    if (at < 0) throw new Error(`expected "key=value", got "${pair}"`)
    const value = pair.slice(at + 1)
    inputs[pair.slice(0, at)] = value.startsWith('@')
      ? new TextDecoder().decode(await readCliFile(inv, value.slice(1), '--field'))
      : value
  }
  return inputs
}

export async function workflowRunCmd(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const id = inv.texts[0] ?? ''
  if (id === '') throw new Error('a workflow ID, name, or filename is required')
  const transport = ghTransport(inv.config)
  const ref = repoFor(inv, fl)
  let branch = fl.asStr('ref') ?? (inv.config as { branch?: string }).branch
  if (branch === undefined || branch === '') {
    const repository = (await viewRepo(transport, ref)) as { default_branch?: unknown }
    branch = typeof repository.default_branch === 'string' ? repository.default_branch : undefined
  }
  if (branch === undefined || branch === '') throw new Error('a workflow ref is required')
  await dispatchWorkflow(transport, ref, id, { ref: branch, inputs: await workflowInputs(inv, fl) })
  return textOut('')
}
